h/services/annotation_moderation.py

```diff
@@ -23,27 +23,60 @@ def all_hidden(self, annotation_ids: str) -> set[str]:
         return {a.id for a in query if a.is_hidden}
 
     def set_status(
-        self, annotation: Annotation, user: User, status: ModerationStatus | None
+        self,
+        annotation: Annotation,
+        status: ModerationStatus | None,
+        user: User | None = None,
     ) -> None:
         """Set the moderation status for an annotation."""
+
+        if not user:
+            # If no user is provided assume this is an automatic transition triggered by the annotation author
+            user = self._user_service.fetch(annotation.userid)
+
         if status and status != annotation.moderation_status:
             self._session.add(
                 ModerationLog(
-                    annotation_id=annotation.id,
+                    annotation=annotation,
                     old_moderation_status=annotation.moderation_status,
                     new_moderation_status=status.value,
-                    moderator_id=user.id,
+                    moderator=user,
                 )
             )
             annotation.moderation_status = status
 
-    def update_status(self, annotation: Annotation) -> None:
+    def update_status(self, action: AnnotationAction, annotation: Annotation) -> None:
+        """Change the moderation status of an annotation based on the action taken."""
         if not annotation.moderation_status and annotation.shared:
             # If an annotation is not private but doesn't have a moderation status
             # it means that the moderation status hasn't been migrated yet.
             # Set the default `APPROVED` status
             annotation.moderation_status = ModerationStatus.APPROVED
 
+        if not annotation.shared:
+            return
+
+        new_status = None
+        pre_moderated = annotation.group.pre_moderated
+        if action == "create":
+            if pre_moderated:
+                new_status = ModerationStatus.PENDING
+            else:
+                new_status = ModerationStatus.APPROVED
+        elif action == "update":
+            if (
+                pre_moderated
+                and annotation.moderation_status == ModerationStatus.APPROVED
+            ):
+                new_status = ModerationStatus.PENDING
+
+            if annotation.moderation_status == ModerationStatus.DENIED:
+                new_status = ModerationStatus.PENDING
```

Comment:
I think there's one behaviour change here for existing users (even though the flag to enabled pre-moderation in a group hasn't been released yet): `DENIED` annotations (that were denied with the <kbd>Hide</kbd> button) will change to `PENDING` if edited by the author. `PENDING` annotations are still hidden, so that's probably okay?

Reply:
Yeah, I think it's fine. The visible behaviour is the same, the anno remains hidden.

If something wrong with all of this we still have the moderation log to revert any of these.